File: devel/cmd/old2new_command.cpp
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: University of Washington <https://www.washington.edu>
// SPDX-FileContributor: 2014-23 Bradley M. Bell
// ----------------------------------------------------------------------------

# include <dismod_at/old2new_command.hpp>
# include <dismod_at/exec_sql_cmd.hpp>
# include <dismod_at/error_exit.hpp>
# include <dismod_at/is_column_in_table.hpp>

namespace dismod_at { // BEGIN_DISMOD_AT_NAMESPACE
/*
-----------------------------------------------------------------------------
{xrst_begin old2new_command}

Convert an Old Database to New Format
#####################################

Syntax
******
``dismod_at`` *database* ``old2new``

data Table
**********
If the original data table does not contain a sample_size column,
this column is added with the value null for each row.
It is an error to run this command on a database that already has
a sample_size column in the data table.

Example
*******
{xrst_toc_list
   example/get_started/old2new_command.py
}

{xrst_end old2new_command}
*/

// ----------------------------------------------------------------------------
void old2new_command(sqlite3* db)
{  //
   // msg
   std::string msg;
   //
   // check if sample_name column is in data table
   std::string table_name  = "data";
   std::string column_name = "sample_size";
   if( dismod_at::is_column_in_table(db, table_name, column_name) )
   {  msg = "old2new commnad: sample_name is alread in data table";
      error_exit(msg, table_name);
   }
   //
   // add sample_name column to data table
   std::string sql_cmd = "ALTER TABLE data ADD COLUMN sample_size;";
   exec_sql_cmd(db, sql_cmd);
   //
   assert( dismod_at::is_column_in_table(db, table_name, column_name) );
   //
   return;
}

} // END_DISMOD_AT_NAMESPACE

File: example/get_started/old2new_command.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# SPDX-FileCopyrightText: University of Washington <https://www.washington.edu>
# SPDX-FileContributor: 2014-23 Bradley M. Bell
# ----------------------------------------------------------------------------
# {xrst_begin old2new_command.py}
# {xrst_comment_ch #}
#
# old2new Command: Example and Test
# #################################
#
# {xrst_literal
#     BEGIN PYTHON
#     END PYTHON
# }
#
# {xrst_end old2new_command.py}
# ---------------------------------------------------------------------------
# BEGIN PYTHON
import sys
import os
import copy
# ---------------------------------------------------------------------------
# check execution is from distribution directory
example = 'example/get_started/old2new_command.py'
if sys.argv[0] != example  or len(sys.argv) != 1 :
   usage  = 'python3 ' + example + '\n'
   usage += 'where python3 is the python 3 program on your system\n'
   usage += 'and working directory is the dismod_at distribution directory\n'
   sys.exit(usage)
#
# import dismod_at
local_dir = os.getcwd() + '/python'
if( os.path.isdir( local_dir + '/dismod_at' ) ) :
   sys.path.insert(0, local_dir)
import dismod_at
#
# import get_started_db example
sys.path.append( os.getcwd() + '/example/get_started' )
import get_started_db
#
# change into the build/example/get_started directory
if not os.path.exists('build/example/get_started') :
   os.makedirs('build/example/get_started')
os.chdir('build/example/get_started')
# ---------------------------------------------------------------------------
# create get_started.db
get_started_db.get_started_db()
# -----------------------------------------------------------------------
# connection
file_name      = 'get_started.db'
connection     = dismod_at.create_connection(
   file_name, new = False, readonly = False
)
#
# data table
sql_cmd = 'ALTER TABLE data DROP COLUMN sample_size;'
dismod_at.sql_command(connection, sql_cmd)
#
# connection
connection.close()
#
# old2new command
program   = '../../devel/dismod_at'
file_name = 'get_started.db'
command   = 'old2new'
dismod_at.system_command_prc( [ program, file_name, command ] )
#
# data_table
file_name      = 'get_started.db'
connection     = dismod_at.create_connection(
   file_name, new = False, readonly = False
)
data_table = dismod_at.get_table_dict(connection, 'data')
connection.close()
#
# check sample_size column
for row in data_table :
   assert row['sample_size'] == None
#
# -----------------------------------------------------------------------
print('old2new_command: OK')
sys.exit(0)
# END PYTHON
